soci/soci_index.go

```diff
@@ -448,10 +443,11 @@ func (b *IndexBuilder) buildSociLayer(ctx context.Context, desc ocispec.Descript
 }
 
 // NewIndex returns a new index.
-func NewIndex(blobs []ocispec.Descriptor, subject *ocispec.Descriptor, annotations map[string]string, legacy bool) *Index {
-	mediaType := OCIArtifactManifestMediaType
-	if legacy {
-		mediaType = ocispec.MediaTypeImageManifest
+func NewIndex(blobs []ocispec.Descriptor, subject *ocispec.Descriptor, annotations map[string]string, artifact bool) *Index {
```

Comment:
Sorry, I was being lazy when I first introduced the `legacy` bool to this. We really should have some sort of `newIndexOptions` struct with a functional arg here.

I think there is a reasonably likely case that artifacts never make it into the OCI image spec in which case it would be pretty unfortunate to be stuck with a useless boolean in the function signature. 

If we make it a functional arg, then if artifacts get removed, we could remove just that function and the only code that would need to change is the code where people were using artifacts. If we keep it a bool, then everyone would need to update code to remove the bool.

Reply:
Since we already have a buildConfig struct would it make sense to change it to have a `manifestType` attribute instead of an `artifact` bool. Something like 
```
type buildConfig struct {
        ...
	artifactsDb        *ArtifactsDb
	platform           ocispec.Platform
	manifestType       string
}
```
We can set the `manifestType` to be either `ocispec.MediaTypeImageManifest` or `ocispec.MediaTypeArtifactManifest`. That way when creating the Index we can set its `MediaType` field directly to `config.ManifestType`

Reply:
Yeah, I think that could make the index builder structure a bit cleaner. The question is how do we pass that to `soci.NewIndex`? Right now it's a bool. We could switch it to a string which is the media type. I think the risk is that if artifacts don't land in the spec, then we have this string where you just always pass the image manifest string which is weird.

Reply:
I think part of the reason behind have the `--manifest-type` argument was to make the CLI extensible/support any new manifest types that may exist in the future. I do think functional arguments would make the function signature a little cleaner but we should remove the `artifact` bool attribute from `buildConfig` if we intend to have a `newIndexOptions` struct since having functionally similar attributes for both would make things repetitive.

Reply:
I think that sounds good, but it's not totally clear to me how we would reduce the repetition. We need some way to pass the info from CLI -> Index builder and from Index builder -> `soci.NewIndex`. Ideally in a way that you only have to pass anything if you're using non-default behavior. I think that would require 2 sets of functional args, but I'm open to reviewing a PR if you have an idea for making it just one set.

Reply:
Yeah I think you are right... I cant think of a way of separating the `bool `logic from build config and index config while maintaining a single functional arg.